drivers/include/matrix_keypad.h

```diff
@@ -0,0 +1,218 @@
+#include <stdbool.h>
+#include "periph/gpio.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Maximum number of rows
+ */
+#ifndef CONFIG_MATRIX_KEYPAD_NUM_ROWS
+#define CONFIG_MATRIX_KEYPAD_NUM_ROWS   2
+#endif
+
+/**
+ * @brief Maximum number of columns
+ */
+#ifndef CONFIG_MATRIX_KEYPAD_NUM_COLUMNS
+#define CONFIG_MATRIX_KEYPAD_NUM_COLUMNS   2
+#endif
+
+/**
+ * @brief Debounce pattern high to low bits
+ */
+#ifndef CONFIG_MATRIX_KEYPAD_DEBOUNCE_PATTERN_BEGIN
+#define CONFIG_MATRIX_KEYPAD_DEBOUNCE_PATTERN_BEGIN  0xC0
+#endif
+
+/**
+ * @brief Debounce pattern low to high bits
+ */
+#ifndef CONFIG_MATRIX_KEYPAD_DEBOUNCE_PATTERN_END
+#define CONFIG_MATRIX_KEYPAD_DEBOUNCE_PATTERN_END  0x7
+#endif
+
+/**
+ * @brief Use open drain GPIO mode
+ */
+#ifndef CONFIG_MATRIX_KEYPAD_ROWS_USE_OPEN_DRAIN
+#define CONFIG_MATRIX_KEYPAD_ROWS_USE_OPEN_DRAIN  0
+#endif
+
+/**
+ * @brief GPIO mode used for the row GPIOs
+ */
+#if CONFIG_MATRIX_KEYPAD_ROWS_USE_OPEN_DRAIN
+#define MATRIX_KEYPAD_ROWS_GPIO_MODE GPIO_OD_PU
+#else
+#define MATRIX_KEYPAD_ROWS_GPIO_MODE GPIO_OUT
+#endif
+
+/**
+ * @brief Debounce pattern mask
+ */
+#define MATRIX_KEYPAD_DEBOUNCE_MASK  \
+    (CONFIG_MATRIX_KEYPAD_DEBOUNCE_PATTERN_BEGIN | CONFIG_MATRIX_KEYPAD_DEBOUNCE_PATTERN_END)
+
+/**
+ * @brief Type definition for a full row (all columns) state, variable width
+ * depending on the number of columns.
+ */
+#if CONFIG_MATRIX_KEYPAD_NUM_COLUMNS <= 8
+typedef uint8_t matrix_keypad_state_row_t;
+#elif CONFIG_MATRIX_KEYPAD_NUM_COLUMNS <= 16
+typedef uint16_t matrix_keypad_state_row_t;
+#elif CONFIG_MATRIX_KEYPAD_NUM_COLUMNS <= 32
+typedef uint32_t matrix_keypad_state_row_t;
+#elif CONFIG_MATRIX_KEYPAD_NUM_COLUMNS <= 64
+typedef uint64_t matrix_keypad_state_row_t;
+#else
+#error Too many columns on matrix keypad.
+#endif
+
+/**
+ * @brief   Device initialization parameters
+ */
+typedef struct {
+    /**
+     * @brief GPIO pin array for the rows. GPIO_UNDEF is allowed and scanning
+     * is skipped for these.
+     */
+    gpio_t rows[CONFIG_MATRIX_KEYPAD_NUM_ROWS];
+    /**
+     * @brief GPIO pin array for the columns. GPIO_UNDEF is allowed and scanning
+     * is skipped for these.
+     */
+    gpio_t columns[CONFIG_MATRIX_KEYPAD_NUM_COLUMNS];
+    /**
+     * @brief Delay in microseconds between configuring the row gpio and reading
+     * out the column. Can be zero to skip the delay.
+     */
+    uint32_t row2col_delay;
+} matrix_keypad_params_t;
+
+/**
+ * @brief   Callback for key state changes
+ *
+ * @param   arg     callback context
+ * @param   row     Row that changed
+ * @param   column  Column that changed
+ * @param   state   New state of the key, 1 = pressed, 0 = released
+ */
+typedef void (*matrix_keypad_cb_t)(void *arg, size_t row, size_t column, bool state);
+
+/**
+ * @brief   Device descriptor for the driver
+ */
+typedef struct {
+    /**
+     * @brief Device initialization parameters
+     */
+    matrix_keypad_params_t params;
```

Comment:
Why not
```suggestion
    const matrix_keypad_params_t *params;
```

Reply:
I'm fine with either, afaik this is a memory latency vs ram usage tradeoff.

Reply:
Is that measurable?
We are also running the code from flash after all.

Reply:
I really don't remember the specifics.

Reply:
But I'm fine with either here, so just confirm if you prefer it in ROM.

Reply:
It depends on the MCU. E.g. AVRs are clocked so slow so that they do not need to wait for ROM access - it really is the flash that is the bottleneck for CPU clock there. But most other MCUs do implement the logic required to wait on bus accesses. E.g. Cortex M7 have to spend a few cycles waiting for a ROM access, even not every RAM region is accessible in a single CPU cycle. However, Cortex M7 also have quite an elaborate cache system to speed things up.

While I do like a to have an as low latency between the button press and the letter appearing in neovim on the terminal, I am pretty confident that the few CPU cycles spend waiting for the flash (on a cache miss) do not contribute noticeably to that latency.

Reply:
Reworked it as suggested above.